### PR TITLE
[cling] Handle variable templates in DeclUnloader

### PR DESCRIPTION
**Pull request by hahnjo**
Fixes https://github.com/root-project/root/issues/13815

### CODE REVIEW

File: interpreter/cling/lib/Interpreter/DeclUnloader.cpp [hunk 1]
```diff
@@ -345,6 +345,22 @@ namespace {
     }
   };
 
+  template <class EntryType>
+  void removeSpecializationImpl(llvm::FoldingSetVector<EntryType>& Specs,
+                                const EntryType* Entry) {
+    llvm::SmallVector<EntryType*, 4> Filtered;
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DeclUnloader.cpp:351]: "Filtered" doesn't tell me whether it's the gold you want to keep, or the coffee ground, i.e. do you throw away the filtered or are they the ones to survive. IIUC the latter is the case here. So I'd recommend to rename the variable to
```suggestion
    // Remove only Entry from Specs. Keep all others.
    llvm::SmallVector<EntryType*, 4> Keep;
```

File: interpreter/cling/lib/Interpreter/DeclUnloader.cpp [hunk 2]
```diff
@@ -345,6 +345,22 @@ namespace {
...
+    for (auto& Spec : Specs)
+      if (&Spec != Entry)
+        Filtered.push_back(&Spec);
+
+    Specs.clear();
+    for (auto* Spec : Filtered) {
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DeclUnloader.cpp:357]: Is there no `swap`?

**Review by Axel-Naumann — APPROVED** [review]
Very nice, thanks!

**Review by vgvassilev — APPROVED** [review]
Nice!

**Pull request merged**